File: python/btcspv/ser.py
import json

from riemann import tx

from btcspv.types import RelayHeader, SPVProof


def hex_deser(h: str) -> bytes:
    body = h
    if h[:2] == '0x':
        body = h[2:]
    return bytes.fromhex(body)


def hex_ser(b: bytes) -> str:
    return f'0x{b.hex()}'


def dict_from_relay_header(r: RelayHeader) -> dict:
    '''
    Args:
        r (RelayHeader): The RelayHeader to be serialized
    Returns:
        (dict): A dictionary representing the RelayHeader with serialized keys
    '''
    return {
        'raw': hex_ser(r['raw']),
        'hash': hex_ser(r['hash']),
        'height': r['height'],
        'prevhash': hex_ser(r['prevhash']),
        'merkle_root': hex_ser(r['merkle_root']),
    }


def serialize_relay_header(r: RelayHeader) -> str:
    '''
    Args:
        r (RelayHeader): The RelayHeader to be serialized
    Returns:
        (str): A JSON-serialized RelayHeader
    '''
    return json.dumps(dict_from_relay_header(r))


def dict_to_relay_header(d: dict) -> RelayHeader:
    '''
    Args:
        d (dict): The dict with serialized keys to be deserialized
    Returns:
        (RelayHeader): The RelayHeader, a TypedDict with deserialized keys
    '''
    return RelayHeader(
        raw=hex_deser(d['raw']),
        hash=hex_deser(d['hash']),
        height=d['height'],
        prevhash=hex_deser(d['prevhash']),
        merkle_root=hex_deser(d['merkle_root'])
    )


def deserialize_relay_header(s: str) -> RelayHeader:
    '''
    Args:
        s (str): A JSON-serialized RelayHeader
    Returns:
        (RelayHeader): The deserialized RelayHeader
    '''
    return dict_to_relay_header(json.loads(s))


def dict_from_spv_proof(s: SPVProof) -> dict:
    '''
    Args:
        s (SPVProof): The SPVProof to be serialized
    Returns:
        (dict): A dictionary representing the SPVProof with serialized keys
    '''
    return {
        'version': hex_ser(s['version']),
        'vin': hex_ser(s['vin']),
        'vout': hex_ser(s['vout']),
        'locktime': hex_ser(s['locktime']),
        'tx_id': hex_ser(s['tx_id']),
        'index': s['index'],
        'intermediate_nodes': hex_ser(s['intermediate_nodes']),
        'confirming_header': dict_from_relay_header(s['confirming_header'])
    }


def serialize_spv_proof(s: SPVProof) -> str:
    '''
    Args:
        s (SPVProof): The SPVProof to be serialized
    Returns:
        (str): A JSON-serialized SPVProof
    '''
    return json.dumps(dict_from_spv_proof(s))


def dict_to_spv_proof(d: dict) -> SPVProof:
    '''
    Args:
        d (dict): The dict with serialized keys to be deserialized
    Returns:
        (SPVProof): The SPVProof, a TypedDict with deserialized keys
    '''

    t = tx.Tx.from_bytes(
        hex_deser(d['version'])
        + hex_deser(d['vin'])
        + hex_deser(d['vout'])
        + hex_deser(d['locktime'])
    )
    return SPVProof(
        tx=t,
        version=hex_deser(d['version']),
        vin=hex_deser(d['vin']),
        vout=hex_deser(d['vout']),
        locktime=hex_deser(d['locktime']),
        tx_id=hex_deser(d['tx_id']),
        intermediate_nodes=hex_deser(d['intermediate_nodes']),
        index=d['index'],
        confirming_header=dict_to_relay_header(d['confirming_header'])
    )


def deserialize_spv_proof(s: str) -> SPVProof:
    '''
    Args:
        s (str): A JSON-serialized SPVProof
    Returns:
        (SPVProof): The deserialized SPVProof
    '''
    return dict_to_spv_proof(json.loads(s))

File: js/src/ser.js
import * as utils from './utils';

/**
 *
 * Takes a Header deserialized from JSON and returns a new object representing the Header
 *   with deserialized Uint8Array in place of serialzed hex strings
 *
 * @param {Object}    o The Header as an unprcessed object, immediately after derserialization
 * @returns {Object}    The Header with deserialized byte arrays
 */
export function objectToHeader(o) {
  /* eslint-disable camelcase */
  const raw = utils.deserializeHex(o.raw);
  const hash = utils.deserializeHex(o.hash);
  const prevhash = utils.deserializeHex(o.prevhash);
  const merkle_root = utils.deserializeHex(o.merkle_root);
  if (raw.length !== 80) {
    throw new TypeError(`Expected 80 bytes, got ${raw.length} bytes`);
  }
  [hash, prevhash, merkle_root].forEach((e) => {
    if (e.length !== 32) {
      throw new TypeError(`Expected 32 bytes, got ${e.length} bytes`);
    }
  });
  /* eslint-enable camelcase */

  return {
    raw,
    hash,
    height: o.height,
    prevhash,
    merkle_root,
  };
}

/**
 *
 * Deserializes a Header object from a JSON string
 *
 * @param {string}    s The Header serialized as a JSON string
 * @returns {Object}    The Header with deserialized byte arrays
 */
export function deserializeHeader(s) {
  return objectToHeader(JSON.parse(s));
}


/**
 *
 * Takes a Header and serialized each byte array. The result is suitable for JSON serialization
 *
 * @param {Object}    o The Header with deserialized byte arrays
 * @returns {Object}    The Header with byte arrays serialized as hex, suitable for serialization
 */
export function objectFromHeader(h) {
  return {
    raw: utils.serializeHex(h.raw),
    hash: utils.serializeHex(h.hash),
    height: h.height,
    prevhash: utils.serializeHex(h.prevhash),
    merkle_root: utils.serializeHex(h.merkle_root),
  };
}

/**
 *
 * Serializes a Header object to a JSON string
 *
 * @param {Object}    s The Header with deserialized byte arrays
 * @returns {string}    The Header serialized as a JSON string
 */
export function serializeHeader(h) {
  return JSON.stringify(objectFromHeader(h));
}

/**
 *
 * Takes a proof deserialized from JSON and returns a new object representing the proof
 *   with deserialized Uint8Array in place of serialzed hex strings
 *
 * @param {Object}    o The proof as an unprcessed object, immediately after derserialization
 * @returns {Object}    The proof with deserialized byte arrays
 */
export function objectToSPVProof(o) {
  /* eslint-disable camelcase */
  const version = utils.deserializeHex(o.version);
  const vin = utils.deserializeHex(o.vin);
  const vout = utils.deserializeHex(o.vout);
  const locktime = utils.deserializeHex(o.locktime);
  const tx_id = utils.deserializeHex(o.tx_id);
  const intermediate_nodes = utils.deserializeHex(o.intermediate_nodes);

  if (tx_id.length !== 32) {
    throw new TypeError(`Expected 32 bytes, got ${tx_id.length} bytes`);
  }

  return {
    version,
    vin,
    vout,
    locktime,
    tx_id,
    index: o.index,
    intermediate_nodes,
    confirming_header: objectToHeader(o.confirming_header)
  };
  /* eslint-enable camelcase */
}

/**
 *
 * Deserializes a SPVProof object from a JSON string
 *
 * @param {string}    s The SPVProof serialized as a JSON string
 * @returns {Object}    The SPVProof with deserialized byte arrays
 */
export function deserializeSPVProof(s) {
  return objectToSPVProof(JSON.parse(s));
}

/**
 *
 * Takes a SPVProof and serialized each byte array. The result is suitable for JSON serialization
 *
 * @param {Object}    o The SPVProof with deserialized byte arrays
 * @returns {Object}    Object byte arrays serialized as hex, suitable for serialization as JSON
 */
export function objectFromSPVProof(s) {
  return {
    version: utils.serializeHex(s.version),
    vin: utils.serializeHex(s.vin),
    vout: utils.serializeHex(s.vout),
    locktime: utils.serializeHex(s.locktime),
    tx_id: utils.serializeHex(s.tx_id),
    index: s.index,
    intermediate_nodes: utils.serializeHex(s.intermediate_nodes),
    confirming_header: objectFromHeader(s.confirming_header)
  };
}

/**
 *
 * Serializes a SPVProof object to a JSON string
 *
 * @param {Object}    s The SPVProof with deserialized byte arrays
 * @returns {string}    The SPVProof serialized as a JSON string
 */
export function serializeSPVProof(s) {
  return JSON.stringify(objectFromSPVProof(s));
}
